src/BindPattern.jl

```diff
@@ -49,22 +50,45 @@ end
 gensym(base::String) = Base.gensym(base)
 
 #
-# Generate a fresh synthetic variable whose name hints at its purpose.
+# Pretty-printing utilities helpful for displaying the state machine
 #
-function gentemp(p::BoundFetchPattern)
-    error("not implemented: gentemp(::$(typeof(p)))")
+pretty(io::IO, p::BoundPattern, state::BinderState) = pretty(io, p)
+function pretty(io::IO, p::BoundFetchPattern, state::BinderState)
+    temp = get_temp(state, p)
+    pretty(io, temp)
+    print(io, " := ")
+    pretty(io, p)
 end
-function gentemp(p::BoundFetchFieldPattern)
-    gensym(string(simple_name(p.input), ".", p.field_name))
+function pretty(io::IO, p::Union{BoundOrPattern, BoundAndPattern}, state::BinderState)
+    op = (p isa BoundOrPattern) ? "||" : "&&"
+    print(io, "(")
+    first = true
+    for sp in p.subpatterns
+        first || print(io, " ", op, " ")
+        first = false
+        pretty(io, sp, state)
+    end
+    print(io, ")")
 end
-function gentemp(p::BoundFetchIndexPattern)
-    gensym(string(simple_name(p.input), "[", p.index, "]"))
+function pretty(io::IO, s::Symbol)
+    print(io, pretty_name(s))
 end
-function gentemp(p::BoundFetchRangePattern)
-    gensym(string(simple_name(p.input), "[", p.first_index, ":(length-", p.from_end, ")]"))
+function pretty_name(s::Symbol)
+    s = string(s)
+    if startswith(s, "##")
+        string("«", simple_name(s), "»")
+    else
+        s
+    end
 end
-function gentemp(p::BoundFetchLengthPattern)
-    gensym(string("length(", simple_name(p.input), ")"))
+struct FrenchName; s::Symbol; end
+Base.show(io::IO, x::FrenchName) = print(io, pretty_name(x.s))
+function pretty(io::IO, expr::Expr)
+    b = MacroTools.prewalk(MacroTools.rmlines, expr)
+    c = MacroTools.prewalk(MacroTools.unblock, b)
+    print(io, MacroTools.postwalk(c) do var
+        (var isa Symbol) ? FrenchName(var) : var
```

Comment:
it looks like `pretty(io::IO, s::Symbol)` ends up in the same place as `FrenchName` - what does `FrenchName` get us?

Reply:
Since the outside call is `print`, we somehow need the symbols in the middle of the expression to be pretty-printed instead of printed.  So we build a new expression with the symbols replaced by their pretty form.